Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp

```diff
@@ -1299,6 +1298,36 @@ bool GPUConnectionToWebProcess::isAlwaysOnLoggingAllowed() const
     return m_sessionID.isAlwaysOnLoggingAllowed() || m_sharedPreferencesForWebProcess.allowPrivacySensitiveOperationsInNonPersistentDataStores;
 }
 
+#if HAVE(AUDIT_TOKEN)
+std::optional<audit_token_t> GPUConnectionToWebProcess::presentingApplicationAuditToken(WebCore::PageIdentifier pageIdentifier) const
+{
+    auto iterator = m_presentingApplicationAuditTokens.find(pageIdentifier);
+    if (iterator != m_presentingApplicationAuditTokens.end())
+        return iterator->value.auditToken();
+
+    if (auto parentAuditToken = protectedGPUProcess()->protectedParentProcessConnection()->getAuditToken())
+        return *parentAuditToken;
+
+    return std::nullopt;
+}
+
+ProcessID GPUConnectionToWebProcess::presentingApplicationPID(WebCore::PageIdentifier pageIdentifier) const
+{
+    if (auto auditToken = presentingApplicationAuditToken(pageIdentifier))
+        return pidForAuditToken(*auditToken);
+
+    return { };
+}
+
+void GPUConnectionToWebProcess::setPresentingApplicationAuditToken(WebCore::PageIdentifier pageIdentifier, const std::optional<CoreIPCAuditToken>& auditToken)
+{
+    if (auditToken)
+        m_presentingApplicationAuditTokens.set(pageIdentifier, *auditToken);
+    else
+        m_presentingApplicationAuditTokens.remove(pageIdentifier);
```

Comment:
Do we need this removal?
Should we simplify and pass a non optional token?

Reply:
I added it so that `GPUConnectionToWebProcess` doesn't accumulate audit tokens for closed pages, but maybe that concern isn't worth the extra complexity given we mostly have a process per page.